### ISSUE TITLE
Fields are referenced globally by their name is the JSON specification file

### ISSUE THREAD

**Issue by JulienBedel** [Orange-Cyberdefense/bof#11]
### TL;DR

When BOF is working with items of the JSON specification file, they are referenced globally by their name. This happens when passing default parameters to a frame, as well as when looking for a dependency match..
As a consequence, using the same field name twice will most likely have an unexpected behaviour (or not work at all). 
This is needed if we want to use "general-purpose" blocks (typically to define complex types such as a string) that are going to appear multiple times in the JSON.

### Description

**Context**

I encountered this problem while implementing OPC UA protocol into BOF.
Following discussion with @claire-lex we concluded that this issue was part of BOF core, evenly affecting KNX implementation.

**Problem**

Following OPC UA protocol specifications, I wanted to define "general-purpose" blocks.
What is meant by "general-purpose" is that instead of being "frame-specific blocks that usually appears once in a frame (ex : "SECURITY_HEADER"), they represent generic types that will much likely be used multiple times in a frame (ex : "STRING").

When BOF is working with items of the JSON specification file, they are referenced globally by their name.
As a consequence, using the same field name twice will most likely have an unexpected behavior (or not work at all).
Therefore, we can't use our "general-purpose" blocks in dependencies, while sometimes needed.

**Example 1**

Let's illustrate the problem with the particular case I encountered.
I wanted to define a generic `NODE_ID` type (see *OPC 10000-6 - UA Specification Part 6 - Mappings 1.04, 5.2.2.9* for an exhaustive definition).

A `NODE_ID` block is composed of two fields :
- `node_id_data_encoding` that specifies how the node will encode its data (typically on 2 or 4 bytes).
- `node_id_data` that is a variable block, depending on the encoding.

Here is an intuitive definition I first wrote in the JSON specification file :

```json
"NODE_ID": [
    {"name": "node_id_data_encoding", "type": "field", "size": 1},
    {"name": "node_id_data", "type": "depends:node_id_data_encoding"}
],
"TWO_BYTES_NODE_ID": [ 
    {"name": "node_id_value", "type": "field", "size": 1}
],
"FOUR_BYTES_NODE_ID": [
    {"name": "node_id_namespace", "type": "field", "size": 1},
    {"name": "node_id_value", "type": "field", "size": 2}
]
```

Depending on the value of `node_id_data_encoding`, the appropriate node format is used.
The association between the encoding byte `node_id_data_encoding`  and the appropriate èsub-block is made in the code section :

```json
"codes" : {
    ...
    "node_id_data_encoding": {
        "00": "TWO_BYTES_NODE_ID",
        "01": "FOUR_BYTES_NODE_ID"
    },
    ...
}
```

Once defined, this "general-purpose" block is used in our "frame-specific" blocks :

```json
"REQUEST_HEADER": [
    {"name": "authentication_token", "type": "NODE_ID"},
    {"name": "timestamp", "type": "field", "size": 8},
    {"name": "request_handle", "type": "field", "size": 4},
    ...
]
```

```json
"OPN_data": [
    {"name": "type_id", "type": "NODE_ID"},
    {"name": "service", "type": "field:depends"},
    ...
],
```

In practice, `authentication_token` will be a `TWO_BYTES_NODE_ID`, and `service` will be a `FOUR_BYTES_NODE_ID`.

The problem appears when we create an empty block and want to fill its values by hand.
How does a `node_id_data` knows its format if it can't find a `node_id_data_encoding` ? We could have passed the `node_id_data_encoding` as a default value on frame creation, but because there are 2 fields with this name, BOF won't be able to find them.

Note that the problem is not happening when filling a frame with raw-bytes, because dependencies will be read on-the-fly, directly on the parent. However raw-bytes frame creation is still affected by this issue, see Example 2 below.

**Example 2**

We can also imagine the following scenario. It illustrates a slightly different problem (but is most probably affecting the same core mechanisms).

```json
"EXAMPLE_BLOCK": [
    {"name": "node_id_1", "type": "NODE_ID"},
    {"name": "node_id_2", "type": "NODE_ID"},
    {"name": "dependent_field", "type": "depends:node_id_value"}
]
```

In the current BOF state, `depends:node_id_value` is going to look for the `node_id_value` in the first encountered node `node_id_1`.
What if it should be looked after in `node_id_2` ?

Note that contrary to Example 1, this issue affects crafted frames as well as frames read from raw bytes.

### Solutions to investigate

**0. Making every block unique ("naive" solution)**

The solution I found for now is to make any "general" block unique, so that we won't have to use default parameters at all.

Solution for Example 1 :

```json
"AUTH_TOKEN_NODE_ID": [
    {"name": "auth_token_node_id_data_encoding", "type": "field", "size": 1},
    {"name": "auth_token_node_id_value", "type": "field", "size": 1}
]
```
```json
"SERVICE_NODE_ID": [
    {"name": "service_node_id_data_encoding", "type": "field", "size": 1},
    {"name": "service_node_id_namespace", "type": "field", "size": 1},
    {"name": "service_node_id_value", "type": "field", "size": 2}
]
```

And are used accordingly in our `authentication_token` and `service`.

Still, there are some potential problems with this solution :

- The "general-purpose" block are not so general now because they must be redefined whenever they are used.
- This redefinition causes some unnecessary repetition in JSON specification file.
- We are getting farther and farther away from the protocol original specifications, altering them to match BOF.
- If some "general-purpose block" block is used inside another one  (typically an OPC UA _ExpandedNodeId_ has 2 variations + a _NodeId_ inside it), we have to write every combination possible. For example here _TwoByteExpandedNodeIdVariation1_, _TwoByteExpandedNodeIdVariation2_, _FourByteExpandedNodeIdVariation1_, _FourByteExpandedNodeIdVariation2_.. instead of _ExpandedNodeId_ containing its two variations according to parameters, and a _NodeId_ depending on parameters too.

**1. Frame default parameters system improvement**

An other solution would be to pass default parameter at their exact position when creating a frame.

Solution for example 1 :

```python
frame = opcua.OpcuaFrame(type='OPN', defaults={'frame.body.request_header.authentication_token.node_id_data_encoding': '00', 'frame.body.opn_data.type_id.node_id_data_encoding': '01'})
```

Still, we have the following problems :

- We need to pass parameters for generic types on frame creation, while they could be specified in the JSON by themself. In fact, the node encoding is dependent on the particular position of the field in the JSON structure, and ideally would not have to be defined by hand.
- This does not solve Example 2.

**2. JSON parent-specified default parameters**

In order to extend the previous solution, we could use the same principle directly into the JSON file.

Dependencies can be specified by their "containing item" :

```json
"REQUEST_HEADER": [
    {"name": "authentication_token", "type": "NODE_ID", "default": "node_id_data_encoding=00"},
    {"name": "timestamp", "type": "field", "size": 8},
    {"name": "request_handle", "type": "field", "size": 4},
]
```

Or maybe an easier way would be to define default values for whole blocks :

```json
"REQUEST_HEADER": [
    {"name": "authentication_token", "type": "NODE_ID", "default": "\\x00\\x01"},
    {"name": "timestamp", "type": "field", "size": 8},
    {"name": "request_handle", "type": "field", "size": 4},
]
```

This still does not solves Example 2, but combined with the previous solution 1. allows to solve the frame creation issue.

**3. JSON child-specified default values**

Instead of having the parent detailing precisely what values should take its child fields. We could take the opposite approach and have the child specify precisely where to look for its default value.

Solution for example 1 :

```json
"NODE_ID": [
    {"name": "node_id_data_encoding", "type": "field", "size": 1, "default": "depends:parent"},
    {"name": "node_id_data", "type": "depends:node_id_data_encoding"}
],
```
And an association code to match the parent name with a specific encoding.

We can't use `depends:frame.request_header.authentication_token.value` because this node is supposed to be generic, therefore the dependent path should be relative and not absolute.

### Help needed

Any help is appreciated, more precisely regarding the following topics :

- Identify if the problem is  general to BOF core, or specific to OPC UA.
- If so, decide if the problem is worth fixing or if we can get around (⇒ trade-off with specification match to think about).
- Identify the appropriate solution (⇒ trade-off with the ease of fix to think about).
- Help fixing the problem.

**Comment by JulienBedel** [comment]
Looking at solution 1, I found two ways to get the absolute path of an item (field or block) in order to use it in dependencies :

1. Add a "path" attribute to fields and blocks, that append their current name to their parent path. Therefore at every depth we know where we are.
2. Add a *Breadth-first search* algorithm (cf. [https://en.wikipedia.org/wiki/Breadth-first_search](https://en.wikipedia.org/wiki/Breadth-first_search)) from the frame when looking for dependency to get all the sub-fields and their path.

**Comment by JulienBedel** [comment]
Hi ! Thanks a lot for providing such a detailed answer, it is really helpful and appreciated !

I did not realize when I left the issue that it was actually referring to multiple topics that should be treated individually. Thanks for pointing that out and put words on each of them !

### TL;DR

- Your message really clarified the issue, decomposing it in sub-topics.
- The solutions you provided are great and I will try to implement them.
- I made a small roadmap to highlight what I will focus on first.

### Details

I agree with every observation you had on the answers. Here are some thoughts on each one :

- Solution 1 : I had not thought that "For some fields, their exact location depends on the frame", I will try to think about it and see what we can do.
- Solution 2 : the example syntax you provided is really clear and would suit the existing BOF defaults value very well !
- Solution 3 : using a syntax to specify levels is essential and will be one of the main focus too.

I eventually agree that a solid basis for every solution is to refer fields by their paths (keeping in mind that they should still be referred by their name only).

### Roadmap

Here is a little roadmap for now :

1. Add parameters to blocks and fields in order to construct a path. I will try to find/use a Python built-in library to handle paths, in order to allow operations such as "../../" without having to redefine them by hand. (After a few minutes of searching the web python's `pathlib`'s *pure paths* might just fit our needs (see : https://docs.python.org/3/library/pathlib.html#pure-paths).

2. Add support for dependency search with absolute and relative path.

3. Add support for default value search with absolute path.

4. Look in details for the solutions implementations considering what has been done previously.

Because my internship (dedicated to the development of BOF, among other things) is ending soon, I will try to focus on providing you basis to address the solution in case I cannot.

If I find new ideas to assess the issue I will make sure to let you know in this thread ;)

**Comment by JulienBedel** [comment]
I added various functionalities affecting dependency search, with the idea of implementing solution 3 based on your feedbacks. New keywords have been introduced, with particular care taken to make them compatible with the previous version of BOF. 

### Addressing Example 1 issue

After some thoughts following your answer, I came up with the following syntax to address Example 1. It is slightly different from what we discussed :

```json
{
    "blocks": {
        "NODE_ID": [
            {"name": "node_id_data_encoding", "type": "field", "size": 1, "default": "depends:relative, attribute:name, level:0"},
            {"name": "node_id_data", "type": "depends:node_id_data_encoding"}
        ],
        "TWO_BYTES_NODE_ID": [ 
            {"name": "node_id_value", "type": "field", "size": 1}
        ],
        "FOUR_BYTES_NODE_ID": [
            {"name": "node_id_namespace", "type": "field", "size": 1},
            {"name": "node_id_value", "type": "field", "size": 2}
        ],
        "REQUEST_HEADER": [
            {"name": "authentication_token", "type": "NODE_ID"}
            ...
        ],
        "OPN_data": [
            {"name": "type_id", "type": "NODE_ID"}
            ...
        ],
    },
    "codes" : {
        "node_id_data_encoding": {
            "00": "TWO_BYTES_NODE_ID",
            "01": "FOUR_BYTES_NODE_ID"
        },
        "node_id_data_encoding_relative": {
            "authentication_token": "00",
            "type_id": "01"
        }
    }
}
```

The main thing to understand here is that `node_id_data_encoding`'s default value depends on the context in which `NODE_ID` is used. In other words, the node needs to know if it is used as an "authentication_token" or as a "type_id". Therefore a node should be able to retrieve the field name it which it is used (here again : "authentication_token" or "type_id").

Let's dive in the new keywords details :

- While it was only possible to look for a field's *value* in BOF. The `attribute:` keyword now enables to to specify the searched item's attribute to look for. For example here, it is the *name* that we want to get. It might not seem usefull on itself because we usually know the field's name, but is necessary when using `depends:relative` because we have no name for our parent block to look in. Note that `attribute:` is defaulted to *value*, as it was typically used in BOF. See details in feature 1 below, commit [efb40390](https://github.com/Orange-Cyberdefense/bof/commit/efb4039061bcd9d995d450972786915d4f28225e).
- `level:` keyword is used to specify the parent's block distance in which looking for the specified item. `level:0` points to self, `level:1` points to parent,`level:2` points to parent's parent, etc..  `level:-1` means to look for the closest matching item in any level, and is passed by default. See details in feature 2 below, commit [87f6cd6](https://github.com/Orange-Cyberdefense/bof/commit/87f6cd6a355c131bd5ac5a5b50dd136d61cfec6a).
- As before, the `depends:`  keyword is used to specify the field we are looking for. In addition, it is possible to use the `depends:relative` value in order to look for the attributes of a parent block, without knowing its name. Because we don't know the parent's name, we use add `_value` to the name of the field containing dependency to build up our codes (here `node_id_data_encoding_relative`). See feature 3 details below, commit [b5622f3a93](https://github.com/Orange-Cyberdefense/bof/commit/b5622f3a93f66214bc162ec6071c56893a859c62).

All those keywords should be compatible with previous JSON syntaxes. For example,  the following syntaxes will be interpreted in the same way :

```json
{"name": "body", "type": "depends:message_type"}
```

```json
{"name": "body", "type": "depends:message_type, keyword:value, level:-1"}
```

> `keyword:value` et `level:-1` are passed by default, resulting in the same behavior as before.

### 1. Dependency search extended to any attribute in a field

Example to set the default value according to another field's size :

```json
"TEST_BLOCK": [
    {"name": "test_field_1", "type": "field", "size": 2},
    {"name": "test_field_2", "type": "field", "size": 1, "default": "depends:test_field_1, attribute:size"}
]
```

This feature is implemented thanks to a new parameter `depends_attribute` (gathered through additional parsing of `attribute:` keyword in `_get_depends`) passed to `get_depends_value`.

This parameter is used to look for fields match as follows :

```python
if depends_value == to_property(field.name):
    if(hasattr(field, depends_attribute)):
        block = self._spec.get_code_value(field.name, getattr(field, depends_attribute))
        return block if block else getattr(field, depends_attribute)
```

> The idea is that compared to previous BOF version `field.value` is replaced with `getattr(field, depends_attribute)`.

### 2. Dependency search at user-specified parent level

Example of dependencies with multiple `test_field` fields in parents blocks. It is now possible to retrieve the field even if it is not the closest one :

```json
"PARENT_2": [
    {"name": "test_field", "type": "field", "size": 1, "default": 2},
    {"name": "test_block_2", "type": "PARENT_1"}
],
"PARENT_1": [
    {"name": "test_field", "type": "field", "size": 1, "default": 1},
    {"name": "test_block_2", "type": "PARENT_0"}
],
"PARENT_0": [
    {"name": "test_field_grandparent", "type": "field", "size": 1, "default": "depends:test_field, level:2"},
    {"name": "test_field_parent", "type": "field", "size": 1, "default": "depends:test_field, level:1"}
],
```

This feature is implemented through a new parameter `level` passed to`get_depends_value`, corresponding to the parsed champ `level:` keyword.

This parameter is then used in order to construct a list of field in which looking for a match with `depends:` :

```python
target_block = None # specifies the block in which we want to look for a dependency match
field_list = None
# if a level is specified, identifies a target block to look into
if isinstance(level, int) and level >= 0:
    current_parent = self
    for i in range (1, level+1):
        if hasattr(current_parent, '_parent') and current_parent._parent:
            current_parent = current_parent._parent
        else:
            raise BOFProgrammingError("Specified level '{0}' is too high, no parent found.".format(level))
    target_block = current_parent
# if a target block is found, look for dependencies in it only
if target_block:
    field_list = list(target_block)
# otherwise do like it was doing before in BOF
else:
    field_list = list(self._parent) + list(self) if self._parent else list(self)
field_list.reverse()
```

### 3. Relative dependency search

In some specific cases, it is important to get information on a parent block without knowing its name. The global example presented as solution to Example 1 illustrates this problem well.

This feature is implemented through the use of the keyword `relative` with `depends:`, in combination with the two new features :

```python
if depends_value == spec.DEPENDS_RELATIVE and hasattr(target_block, depends_attribute):
	block = self._spec.get_code_value(template['name'] + spec.DEPENDS_RELATIVE_CODE_EXTENSION, getattr(target_block, depends_attribute))
    return block if block else getattr(target_block, depends_attribute)
```

We simply check for the presence of the keyword `relative ` (=`spec.DEPENDS_RELATIVE`) in `depends:` value and gather the specified `depends_attribute`.

### Advancement on issue solving

Example 1 has been addressed, and can now be handled by BOF. Implementing :

- Solution 3 : it is possible to retrieve a parameter even if it is not the closest one.

However, different needs you highlighted in your answer have not  been implemented yet :

- Solution 1 : it is not yet possible to refer to the fields by their path (in user-specified values as well as in dependencies). Nevertheless, path attribute that automatically takes its value have been added to fields and blocks (even if they are not used yet).
- Solution 2 : Extend default value to blocks in the JSON file (would solve Example 1 too, I might should have started with that actually...).

It is possible that while focusing on Example 1, I implemented features that are a bit 'off-topic'. I will try to focus on solution 1 and 2 see if they could replace those features. Feel free to tell me what you think of it (if they should be replaced, improved, etc).

### Limitations

- Feature 2 allows to specify targeted parents for dependency search, but won't work with uncles or cousins, which might be needed.
- Feature 3 make us of an hard-coded keyword, there might be another way implement it.
- I used `_parent` parameter without any getter, we might have define one.
- I don't really took the time to set up the test environment for KNX, thereby I just know that the implementation passes OPC UA's tests.

**Comment by claire-lex** [comment]
This issue is not relevant anymore, as our next release no longer relies on JSON files for protocol implementations, but on Scapy.
(Preview in branch `scapybof`)
